### PR TITLE
Hashicorp Vault support

### PR DESCRIPTION
**Pull request by LordGaav**
Adds support for Hashicorp's Vault.

Vault can be used to securely store secrets, which can only be read by certain clients based on ACLs. See the full documentation here: https://www.vaultproject.io/docs/index.html .

### CODE REVIEW

**Comment by LordGaav** [comment]
@uhavin would you mind reviewing this PR?

**Comment by LordGaav** [comment]
Documentation added, ready for review @uhavin

File: omniconf/tests/backends/test_vault.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+            except:  # pragma: nocover
+                cls.manager.stop()
+                raise
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.manager.stop()
+
+    def setUp(self):
+        self.assertTrue(self.root_client.is_authenticated)
```
Comment (LordGaav) [inline omniconf/tests/backends/test_vault.py]: True, but if the root_client is not authenticated there is no point in executing the test itself. I'm guarding against the root_client getting unauthenticated during tests.

I'm open to alternatives however.

Comment (uhavin) [inline omniconf/tests/backends/test_vault.py]: If the behaviour of the root_client is predictable, there should be no need for the assertion. If it is unpredictable, you should definitely use mock.

Comment (LordGaav) [inline omniconf/tests/backends/test_vault.py]: Alright, I'll remove the assertion from setUp.

File: omniconf/tests/backends/test_vault.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+                cls.root_client.token = cls.manager.root_token
+                cls.root_client.set_policy("deny", DENY_POLICY)
+                cls.root_client.set_policy("normal", NORMAL_POLICY)
+                cls.deny_token = cls.root_client.create_token(
+                    id="DENY", policies=["deny"])['auth']['client_token']
+                cls.normal_token = cls.root_client.create_token(
+                    id="NORMAL", policies=["normal"])['auth']['client_token']
+            except:  # pragma: nocover
+                cls.manager.stop()
+                raise
```
Comment (LordGaav) [inline omniconf/tests/backends/test_vault.py]: Done.

**Pull request merged**